corehq/apps/registry/models.py

```diff
@@ -124,3 +142,167 @@ class RegistryPermission(models.Model):
+    domain = models.CharField(max_length=255)
+    user = models.ForeignKey(User, related_name="registry_actions", on_delete=models.CASCADE)
+    related_object_id = models.CharField(max_length=32)
+    related_object_type = models.CharField(max_length=32, choices=RELATED_OBJECT_CHOICES)
+    detail = JSONField(null=True)
+
+
+class RegistryAuditHelper:
+    def __init__(self, registry):
+        self.registry = registry
+
+    def registry_activated(self, user):
+        self.log_registry_activated_deactivated(user, is_activated=True)
+
+    def registry_deactivated(self, user):
+        self.log_registry_activated_deactivated(user, is_activated=False)
+
+    def invitation_accepted(self, user, invitation):
+        return self.log_invitation_accepted_rejected(user, invitation, is_accepted=True)
+
+    def invitation_rejected(self, user, invitation):
+        return self.log_invitation_accepted_rejected(user, invitation, is_accepted=False)
+
+    def invitation_added(self, user, invitation):
+        return self.log_invitation_added_removed(user, invitation, is_added=True)
+
+    def invitation_removed(self, user, invitation):
+        return self.log_invitation_added_removed(user, invitation, is_added=False)
+
+    def grant_added(self, user, grant):
+        return self.log_grant_added_removed(user, grant, is_added=True)
+
+    def grant_removed(self, user, grant):
+        return self.log_grant_added_removed(user, grant, is_added=False)
+
+    def schema_changed(self, user, new, old):
+        return RegistryAuditLog.objects.create(
+            registry=self.registry,
+            user=user,
+            action=RegistryAuditLog.ACTION_SCHEMA_CHANGED,
+            domain=self.registry.domain,
+            related_object_id=self.registry.id,
+            related_object_type=RegistryAuditLog.RELATED_OBJECT_REGISTRY,
+            detail={
+                "new_schema": new,
+                "old_schema": old,
+            }
+        )
+
+    def data_accessed(self, user, domain, related_object, filters=None):
+        if not related_object or not hasattr(related_object, "doc_type"):
+            raise ValueError("Unexpected related object")
+
+        if related_object.doc_type == "ReportConfiguration":
+            related_object_id = related_object.get_id
+            related_object_type = RegistryAuditLog.RELATED_OBJECT_UCR
+        elif related_object.doc_type == "Application":
+            related_object_id = related_object.get_id
+            related_object_type = RegistryAuditLog.RELATED_OBJECT_APPLICATION
+        else:
+            raise ValueError(f"Unexpected related object type: {related_object.doc_type}")
+
+        return RegistryAuditLog.objects.create(
+            registry=self.registry,
+            user=user,
+            action=RegistryAuditLog.ACTION_DATA_ACCESSED,
+            domain=domain,
+            related_object_id=related_object_id,
+            related_object_type=related_object_type,
+            detail=filters
```

Comment:
`filters` might need some check that it's an expected format, but I'm okay deferring that until this sees a use-case.